idea/idea-fir/src/org/jetbrains/kotlin/idea/core/overrideImplement/KtGenerateMembersHandler.kt

```diff
@@ -0,0 +1,284 @@
+                    put(overriddenSymbol.psi ?: continue, node)
+                }
+            }
+        }
+
+        // Note on implementation: here we need the original ordering defined in the source code, so we stick to PSI rather than using
+        // `KtMemberScope` because FIR groups overloaded functions together.
+        outer@ for ((selectedSymbol, generatedPsi) in selectedMemberSymbols) {
+            val superSymbol = selectedSymbol.originalOverriddenSymbol
+            val superPsi = superSymbol?.psi
+            if (superPsi == null) {
+                // This normally should not happen, but we just try to play safe here.
+                sentinelTailNode.prepend(DoublyLinkedNode(MemberEntry.NewEntry(generatedPsi)))
+                continue
+            }
+            var currentPsi = superSymbol.psi?.prevSibling
+            while (currentPsi != null) {
+                val matchedNode = superPsiToMemberEntry[currentPsi]
+                if (matchedNode != null) {
+                    val newNode = DoublyLinkedNode<MemberEntry>(MemberEntry.NewEntry(generatedPsi))
+                    matchedNode.append(newNode)
+                    superPsiToMemberEntry[superPsi] = newNode
+                    continue@outer
+                }
+                currentPsi = currentPsi.prevSibling
+            }
+            currentPsi = superSymbol.psi?.nextSibling
+            while (currentPsi != null) {
+                val matchedNode = superPsiToMemberEntry[currentPsi]
+                if (matchedNode != null) {
+                    val newNode = DoublyLinkedNode<MemberEntry>(MemberEntry.NewEntry(generatedPsi))
+                    matchedNode.prepend(newNode)
+                    superPsiToMemberEntry[superPsi] = newNode
+                    continue@outer
+                }
+                currentPsi = currentPsi.nextSibling
+            }
+            val newNode = DoublyLinkedNode<MemberEntry>(MemberEntry.NewEntry(generatedPsi))
+            superPsiToMemberEntry[superPsi] = newNode
+            sentinelTailNode.prepend(newNode)
+        }
+        return sentinelHeadNode.toListSkippingNulls()
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    private fun insertMembersAccordingToPreferredOrder(
+        symbolsInPreferredOrder: List<MemberEntry>,
+        classLeftBrace: PsiElement?,
+        editor: Editor,
+        currentClass: KtClassOrObject
+    ): List<Iterable<KtDeclaration>> {
+        var firstAnchor: PsiElement? = null
+        if (symbolsInPreferredOrder.first() is MemberEntry.NewEntry) {
+            val firstExistingEntry = symbolsInPreferredOrder.firstIsInstanceOrNull<MemberEntry.ExistingEntry>()
+            if (firstExistingEntry != null) {
+                firstAnchor = firstExistingEntry.psi.prevSiblingOfSameType() ?: classLeftBrace
+            }
+        }
+
+        val entryBatchAndAnchor = mutableListOf<Pair<List<KtCallableDeclaration>, PsiElement?>>()
```

Comment:
Do we really need a list of pairs of lists?

Reply:
Let me use a custom data class instead.